ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartUploadPartHandler.java

```diff
@@ -65,28 +107,202 @@ void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        idConverter.convert(restRequest, blobId, blobInfo, idConverterCallback(blobInfo, blobId));
+      }, uri, logger, finalCallback);
+    }
+
+    /**
+     * After {@link IdConverter#convert} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @param blobInfo the {@link BlobInfo} to use for security checks.
+     * @param blobId the blob ID returned by the router (without decoration or obfuscation by id converter).
+     * @return a {@link Callback} to be used with {@link IdConverter#convert}.
+     */
+    private Callback<String> idConverterCallback(BlobInfo blobInfo, String blobId) {
+      return buildCallback(frontendMetrics.putIdConversionMetrics, convertedBlobId -> {
+        restResponseChannel.setHeader(RestUtils.Headers.LOCATION, convertedBlobId);
+        securityService.processResponse(restRequest, restResponseChannel, blobInfo, securityProcessResponseCallback());
+      }, uri, logger, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#processResponse}, call {@code finalCallback}.
+     * @return a {@link Callback} to be used with {@link SecurityService#processResponse}.
+     */
+    private Callback<Void> securityProcessResponseCallback() {
+      return buildCallback(frontendMetrics.putBlobSecurityProcessResponseMetrics, securityCheckResult -> {
+        try {
+          // Set the response status to 200 since Ambry named blob PUT has response as 201.
+          if (restResponseChannel.getStatus() == ResponseStatus.Created) {
+            restResponseChannel.setStatus(ResponseStatus.Ok);
+          }
+
+          // Set S3 ETag header
+          String blobId = (String) restResponseChannel.getHeader(LOCATION);
+          restResponseChannel.setHeader("ETag", blobId);
+          //TODO [S3]: remove x-ambry- headers
+          finalCallback.onCompletion(null, null);
+        } catch (RestServiceException e) {
+          finalCallback.onCompletion(null, e);
         }
+      }, restRequest.getUri(), logger, finalCallback);
+    }
 
-        // Set S3 ETag header
-        String blobId = (String) restResponseChannel.getHeader(LOCATION);
-        restResponseChannel.setHeader("ETag", blobId);
-        //TODO [S3]: remove x-ambry- headers
-        callback.onCompletion(null, null);
-      } catch (RestServiceException e) {
-        callback.onCompletion(null, e);
+    /**
+     * Parse {@link BlobInfo} from the request arguments. This method will also ensure that the correct account and
+     * container objects are attached to the request.
+     * @return the {@link BlobInfo} parsed from the request arguments.
+     * @throws RestServiceException if there is an error while parsing the {@link BlobInfo} arguments.
+     */
+    private BlobInfo getBlobInfoFromRequest() throws RestServiceException {
```

Comment:
Can this be moved to a util class shared with NamedBlobPutHandler and PostBlobHandler?

Reply:
Yes, it has the potential to consolidate. 
But not sure if we should consolidate with generic NamedBlobPutHandler in case S3 has some special implementation.
I added the TODO. Let's revisit it after we have done with the S3 blob life cycle management.